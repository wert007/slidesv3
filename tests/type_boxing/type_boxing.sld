struct Person {
    name: string;
    age: int;
}

struct Value {
    value: int;

    func $toString() -> string {
        return 'Value(' + this.value + ')';
    }
}

func fib(n: int) -> int {
    if n <= 0 {
        return 1;
    } else if n <= 2 {
        return n;
    } else {
        return fib(n - 1) + fib(n - 2);
    }
}

func printLog(obj: any) {
    print('Log: ' + obj);
}

func main() {
    let log = printLog;
    // let fs = [
    //     print,
    //     printLog
    // ];
    log(42);
    log(true);
    log(none);
    let n : int? = 13;
    log(n);
    n = none;
    log(n);
    log([1, 2, 3,]);
    log([[1], [2], [3],]);
    log(print);
    log('Hello World!');
    log(fib);
    let s : string? = 'Noneable strings work just fine';
    log(s);
    log(new Person('John Doe', 32));
    let v = new Value(42);
    log(v);
    let maybeV : Value? = none;
    log(maybeV);
    maybeV = v;
    log(maybeV);
    log([new Value(1), new Value(2), new Value(3), ]);
}
