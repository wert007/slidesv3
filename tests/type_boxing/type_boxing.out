Log: 42
Log: true
Log: none
Log: 13
Log: none
Log: [ 1, 2, 3 ]
Log: [ [ 1 ], [ 2 ], [ 3 ] ]
Log: system call print
Log: Hello World!
Log: fn fib(int) -> int
Log: Noneable strings work just fine
Log: struct Person {
    name = John Doe,
    age = 32,
}
Log: Value(42)
Log: none
Log: Value(42)
Log: [ Value(1), Value(2), Value(3) ]
0: 0
1: 0
2: 0
3: 0
4: 0
5: 0
= -1
